src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1130,13 +1128,45 @@ CleanUpSessions(DistributedExecution *execution)
 			 */
 			CloseConnection(connection);
 		}
-		else if (transactionState == REMOTE_TRANS_CLEARING_RESULTS)
+		else if (connection->connectionState == MULTI_CONNECTION_CONNECTED)
 		{
-			/* TODO: should we keep this in the state machine? or cancel? */
-			ClearResults(connection, false);
-		}
+			RemoteTransaction *transaction = &(connection->remoteTransaction);
+			RemoteTransactionState transactionState = transaction->transactionState;
 
-		/* TODO: can we be in SENT_COMMAND? */
+			if (transactionState == REMOTE_TRANS_CLEARING_RESULTS)
+			{
+				/*
+				 * We might have established the connection, and even sent BEGIN, but not
+				 * get to the point where we assigned a task to this specific connection
+				 * (because other connections in the pool already finished all the tasks).
+				 */
+				Assert(session->commandsSent == 0);
+
+				ClearResults(connection, false);
+			}
+			else if (!(transactionState == REMOTE_TRANS_INVALID ||
+					   transactionState == REMOTE_TRANS_STARTED))
+			{
+				/*
+				 * We don't have to handle anything else. Note that the execution
+				 * could only finish on connectionStates of MULTI_CONNECTION_CONNECTING,
+				 * MULTI_CONNECTION_FAILED and MULTI_CONNECTION_CONNECTED. The first two
+				 * are already handled above.
+				 *
+				 * When we're on MULTI_CONNECTION_CONNECTED, TransactionStateMachine
+				 * ensures that all the necessary commands are successfully sent over
+				 * the connection and everything is cleared up. Otherwise, we'd have been
+				 * on MULTI_CONNECTION_FAILED state.
+				 */
+				elog(WARNING, "unexpected transaction state at the end of execution: %d",
+					 transactionState);
+			}
+		}
+		else
+		{
+			elog(WARNING, "unexpected connection state at the end of execution: %d",
+				 connection->connectionState);
```

Comment:
I guess we usually handle it immediately, but I'm a little worried that we can let the connection sit in a state like `MULTI_CONNECTION_LOST` and do not notice it during tests.

Reply:
Sorry, I couldn't follow your concern for this. `MULTI_CONNECTION_LOST ` is immediately goes to state `MULTI_CONNECTION_FAILED`, also `MULTI_CONNECTION_INITIAL` goes to `MULTI_CONNECTION_CONNECTED`.

There is one another state `MULTI_CONNECTION_WAIT_RETRY` which is never used in the code and I don't know what is that.  Do you think would that be a problem?

Reply:
I don't think there is a problem here. It's just to make the code more robust. The loop in the StateMachine functions is just an optimisation that I'd rather not rely on.

Some other code path could start using the `MULTI_CONNECTION_LOST` without realising that the adaptive executor is not tolerant to getting a connection in that state from the connection API. Perhaps we already have that possibility somewhere (e.g. an OOM causes a rollback to savepoint), and users get flushed with warnings in production which we never see in testing.

I think we can probably remove `MULTI_CONNECTION_WAIT_RETRY` from the unified_executor branch. It does not exist on master.

Reply:
Then, I guess it is safer to `CloseConnection` for `MULTI_CONNECTION_LOST`, similar to `MULTI_CONNECTION_FAILED` ?